src/core/coap/coap_client.cpp

```diff
@@ -42,16 +43,19 @@
 namespace Thread {
 namespace Coap {
 
-Client::Client(Ip6::Netif &aNetif):
-    mSocket(aNetif.GetIp6().mUdp),
+Client::Client(Ip6::Netif &aNetif, SenderFunction aSender, ReceiverFunction aReceiver):
+    CoapBase(aNetif.GetIp6().mUdp, aSender, aReceiver),
     mRetransmissionTimer(aNetif.GetIp6().mTimerScheduler, &Client::HandleRetransmissionTimer, this)
 {
     mMessageId = static_cast<uint16_t>(otPlatRandomGet());
 }
 
 ThreadError Client::Start()
 {
-    return mSocket.Open(&Client::HandleUdpReceive, this);
+    Ip6::SockAddr addr;
+    addr.mPort = static_cast<Ip6::Udp *>(mSocket.mTransport)->GetEphemeralPort();
+
+    return CoapBase::Start(addr);
```

Comment:
The Joiner needs to register communication port as unsecured before initiating DTLS handshake after Discovery, so we must set the port earlier, at least for the secure client. But I didn't want do differentiate this behavior between secure/regular client, hence I set it in the base class.